### ISSUE TITLE
Symbolic script produces wrong gradients

### ISSUE THREAD

**Issue by ngimel** [pytorch/pytorch#66883]
## 🐛 Bug

When symbolic script creates a DifferentiableGraph where a variable is used inside the graph, and also is output of the graph and used in subsequent operations, the contribution from subsequent operations is not added to the gradient. 
## To Reproduce

1. Enable layer_norm in symbolic script, recompile pytorch: 
```
diff --git a/torch/csrc/jit/runtime/symbolic_script.cpp b/torch/csrc/jit/runtime/symbolic_script.cpp
index 8a0266e7db..b0fbe60ba6 100644
--- a/torch/csrc/jit/runtime/symbolic_script.cpp
+++ b/torch/csrc/jit/runtime/symbolic_script.cpp
@@ -1141,7 +1141,7 @@ const std::vector<std::string> functions = {
 
             return output, backward
 
-        def layer_norm_disabled(input : Tensor,
+        def layer_norm(input : Tensor,
                        normalized_shape : List[int],
```
2. Run the following [script](https://gist.github.com/ngimel/b37141636bdee3ce1a94c4dc0204a24b). The output will be something like 
```
eager model
Starting iterations 
=== iteration: 0, loss: 4.378854274749756 ===
backward
end backward
=== iteration: 1, loss: 2.726267099380493 ===
backward
end backward
script model
Starting iterations 
=== iteration: 0, loss: 4.378854274749756 ===
backward
end backward
=== iteration: 1, loss: 10.749004364013672 ===
backward
end backward
```
Note how even though model is deterministic, the loss of the second iteration is very different in the scripted mode than in the eager mode. Also note that to get these results you need to disable profiling executor, otherwise it doesn't create DifferentiableSubgraph. You can set `debug=True` in the last line to get more info. 

## Expected behavior

Symbolic script should produce correct gradients
## Environment

Fresh master build

## Additional Context

The problem is with the DifferentiableGraph created by symbolic_script (run with debug=True to get a print of DifferentiableGraph). The relevant part of the model is as follows:
```
        norm = self.layer_norm(dense_out)
        split_sum = self.linear_split_1(norm) #* 2 + 1
        prod = self.linear(split_sum) * norm #* 2 + 1
        interaction = torch.sum(torch.stack([norm, prod], dim=0), dim=0)
```      
Note how `norm`, which is `layer_norm` output, participates in `linear_split_1` (where it is split into 3 parts and other things are done to it), `prod`, where split_sum is multiplied by `norm` again, and `interaction`. That means that `dY` for `layer_norm` should have contributions from all these 3 functions. 
DifferentialGraph0 is as follows:
```
with prim::DifferentiableGraph_0 = graph(%weight.28 : Float(*, requires_grad=0, device=cpu),
      %bias.28 : Float(*, requires_grad=0, device=cpu),
      %input.6 : Float(*, *, requires_grad=0, device=cpu),
      %weight.30 : Float(*, *, requires_grad=0, device=cpu),
      %bias.30 : Float(*, requires_grad=0, device=cpu)):
  %1 : int[] = prim::Constant[value=[85, 85, 86]]()
  %8 : float = prim::Constant[value=1.0000000000000001e-05]()
  %5 : int[] = prim::Constant[value=[256]]()
  %69 : int = prim::Constant[value=1]() # <string>:76:118
  %16 : Tensor(requires_grad=0) = aten::linear(%input.6, %weight.30, %bias.30), scope: __module.dense_arch # /data/users/ngimel/pytorch/torch/nn/functional.py:1896:0
  %input.9 : Tensor(requires_grad=0) = aten::relu(%16) # /data/users/ngimel/pytorch/../playground/autodiff_layer_norm.py:108:0
  %output.1 : Tensor(requires_grad=0), %mean.4 : Tensor, %rstd.4 : Tensor = aten::native_layer_norm(%input.9, %5, %weight.28, %bias.28, %8) # <string>:33:33
  %22 : Tensor[] = aten::split_with_sizes(%output.1, %1, %69) # <string>:128:19
  return (%22, %output.1, %input.9, %mean.4, %rstd.4)
  ```
 It returns raw output of `layer_norm` as `output.1`, and it seems that contributions from prod and interaction should flow to `gradOutput` of `output.1` and then be added to the `gradInput` computed by `split_with_sizes` backward (which is `cat`) before being passed to `layer_norm_backward`. Yet that's not what's happening in script, I'm printing means of cat inputs and layer_norm_backward `dY`:
 ```
0 tensor mean -0.000689346 size [32, 85] stride [85, 1]
1 tensor mean 0.00053283 size [32, 85] stride [85, 1]
2 tensor mean 0.000168846 size [32, 86] stride [86, 1]
cat serial result 4.75354e-06 [32, 256][256, 1]
layer norm dY 4.75354e-06 input 0.139588 weight 1
```
cat output directly flows to layer_norm_backward as dY!
In non-scripted mode, that's not the case, and layer_norm dY is different from cat output:
```
0 tensor mean -0.000689346 size [32, 85] stride [85, 1]
1 tensor mean 0.00053283 size [32, 85] stride [85, 1]
2 tensor mean 0.000168846 size [32, 86] stride [86, 1]
cat serial result 4.75351e-06 [32, 256][256, 1]
layer norm dY 0.000410447 input 0.139588 weight 1
```
Even though this particular bug is triggered only with legacy executor, it's a genuine bug in autodiff construction, and if profiling executor was able to construct this DifferentiableGraph, it would hit the same.

**Comment by eellison** [comment]
@Krovatkin landed changes so this no longer fails.
